File: threading/python/13_h2o_molecule.py
"""
QUESTION 13: Building H2O (LeetCode 1117 - Databricks)
======================================================

Problem: Multiple threads call hydrogen() and oxygen().
Ensure they proceed in groups of 2 hydrogen + 1 oxygen.

Key Concepts: Thread synchronization, Barrier patterns, Resource matching
"""

import threading
import time
import random


class H2O:
    """
    H2O Molecule Builder.
    
    EXPLANATION:
    Use semaphores and barriers to synchronize:
    - Hydrogen semaphore: Allows 2 hydrogen threads
    - Oxygen semaphore: Allows 1 oxygen thread
    - Barrier: All 3 must arrive before proceeding
    - After barrier, reset semaphores for next molecule
    """
    
    def __init__(self):
        self.h_semaphore = threading.Semaphore(2)  # 2 H allowed
        self.o_semaphore = threading.Semaphore(1)  # 1 O allowed
        self.barrier = threading.Barrier(3)  # 3 threads to form H2O
        self._lock = threading.Lock()
    
    def hydrogen(self, release_hydrogen):
        self.h_semaphore.acquire()  # Wait for H slot
        
        release_hydrogen()  # Output 'H'
        
        self.barrier.wait()  # Wait for 2H + 1O
        
        # Reset for next molecule
        self.h_semaphore.release()
    
    def oxygen(self, release_oxygen):
        self.o_semaphore.acquire()  # Wait for O slot
        
        release_oxygen()  # Output 'O'
        
        self.barrier.wait()  # Wait for 2H + 1O
        
        # Reset for next molecule
        self.o_semaphore.release()


class H2OAlternative:
    """
    Alternative using Condition Variables.
    
    EXPLANATION:
    Track counts and use conditions:
    - H threads wait if 2 H already waiting
    - O threads wait if 1 O already waiting
    - Form molecule when 2H + 1O ready
    """
    
    def __init__(self):
        self.h_count = 0
        self.o_count = 0
        self.condition = threading.Condition()
    
    def hydrogen(self, release_hydrogen):
        with self.condition:
            while self.h_count == 2:
                self.condition.wait()
            self.h_count += 1
            release_hydrogen()
            self._try_form_molecule()
    
    def oxygen(self, release_oxygen):
        with self.condition:
            while self.o_count == 1:
                self.condition.wait()
            self.o_count += 1
            release_oxygen()
            self._try_form_molecule()
    
    def _try_form_molecule(self):
        if self.h_count == 2 and self.o_count == 1:
            self.h_count = 0
            self.o_count = 0
            self.condition.notify_all()


def test_h2o():
    h2o = H2O()
    output = []
    lock = threading.Lock()
    
    def release_h():
        with lock:
            output.append('H')
    
    def release_o():
        with lock:
            output.append('O')
    
    # Create threads: 4H + 2O = 2 molecules
    threads = []
    for _ in range(4):
        threads.append(threading.Thread(target=h2o.hydrogen, args=(release_h,)))
    for _ in range(2):
        threads.append(threading.Thread(target=h2o.oxygen, args=(release_o,)))
    
    random.shuffle(threads)  # Random order
    
    for t in threads:
        t.start()
    for t in threads:
        t.join()
    
    result = ''.join(output)
    print(f"Output: {result}")
    
    # Verify: Should have valid H2O molecules
    # Count H and O in each group of 3
    for i in range(0, len(result), 3):
        group = sorted(result[i:i+3])
        assert group == ['H', 'H', 'O'], f"Invalid molecule: {result[i:i+3]}"
    
    print("✓ All molecules valid!")


if __name__ == "__main__":
    test_h2o()

File: threading/cpp/13_h2o_molecule.cpp
/*
 * QUESTION 13: Building H2O (LeetCode 1117 - Databricks)
 * ======================================================
 * 
 * Problem: Multiple threads call hydrogen() and oxygen().
 * Ensure they proceed in groups of 2 hydrogen + 1 oxygen.
 * 
 * Key Concepts: Barrier patterns, semaphores, resource matching
 * 
 * Compile: g++ -std=c++20 -pthread 13_h2o_molecule.cpp -o h2o
 */

#include <iostream>
#include <thread>
#include <mutex>
#include <condition_variable>
#include <semaphore>
#include <barrier>
#include <functional>
#include <vector>
#include <random>

// SOLUTION 1: Semaphores + Barrier
class H2O_Sem {
    /*
     * EXPLANATION:
     * - h_sem: Allows 2 hydrogen threads
     * - o_sem: Allows 1 oxygen thread
     * - barrier: All 3 must arrive before proceeding
     * - After barrier, semaphores released for next molecule
     */
private:
    std::counting_semaphore<2> h_sem_{2};
    std::counting_semaphore<1> o_sem_{1};
    std::barrier<> barrier_{3};

public:
    void hydrogen(std::function<void()> releaseHydrogen) {
        h_sem_.acquire();
        releaseHydrogen();
        barrier_.arrive_and_wait();
        h_sem_.release();
    }

    void oxygen(std::function<void()> releaseOxygen) {
        o_sem_.acquire();
        releaseOxygen();
        barrier_.arrive_and_wait();
        o_sem_.release();
    }
};

// SOLUTION 2: Condition Variables
class H2O_CV {
    /*
     * EXPLANATION:
     * Track counts and use condition to synchronize.
     * Form molecule when 2H + 1O ready.
     */
private:
    std::mutex mutex_;
    std::condition_variable cv_;
    int h_count_ = 0;
    int o_count_ = 0;

    void tryFormMolecule() {
        if (h_count_ >= 2 && o_count_ >= 1) {
            h_count_ -= 2;
            o_count_ -= 1;
            cv_.notify_all();
        }
    }

public:
    void hydrogen(std::function<void()> releaseHydrogen) {
        std::unique_lock<std::mutex> lock(mutex_);
        cv_.wait(lock, [this] { return h_count_ < 2; });
        ++h_count_;
        releaseHydrogen();
        tryFormMolecule();
    }

    void oxygen(std::function<void()> releaseOxygen) {
        std::unique_lock<std::mutex> lock(mutex_);
        cv_.wait(lock, [this] { return o_count_ < 1; });
        ++o_count_;
        releaseOxygen();
        tryFormMolecule();
    }
};

void test_h2o() {
    H2O_Sem h2o;
    std::string output;
    std::mutex m;
    
    auto releaseH = [&]() { std::lock_guard<std::mutex> l(m); output += 'H'; };
    auto releaseO = [&]() { std::lock_guard<std::mutex> l(m); output += 'O'; };
    
    std::vector<std::thread> threads;
    
    // 4H + 2O = 2 molecules
    for (int i = 0; i < 4; ++i) {
        threads.emplace_back([&]() { h2o.hydrogen(releaseH); });
    }
    for (int i = 0; i < 2; ++i) {
        threads.emplace_back([&]() { h2o.oxygen(releaseO); });
    }
    
    // Shuffle thread start order
    std::random_device rd;
    std::mt19937 g(rd());
    std::shuffle(threads.begin(), threads.end(), g);
    
    for (auto& t : threads) {
        if (t.joinable()) t.join();
    }
    
    std::cout << "Output: " << output << std::endl;
    
    // Verify: Each group of 3 should be HHO
    bool valid = true;
    for (size_t i = 0; i < output.size(); i += 3) {
        std::string group = output.substr(i, 3);
        std::sort(group.begin(), group.end());
        if (group != "HHO") valid = false;
    }
    
    std::cout << (valid ? "✓ Valid H2O molecules!" : "✗ Invalid!") << std::endl;
}

int main() {
    test_h2o();
    return 0;
}

